From f8b4ff0764be38b9bfaa7b4f891a4497d4726030 Mon Sep 17 00:00:00 2001
From: alandiegosantos <alandiegosantos@gmail.com>
Date: Mon, 16 May 2022 20:21:59 +0200
Subject: EnvoyFilter is also applied to HTTP Proxy Listener (#38924)

The function patchOneListener, that applies the EnvoyFilter patches in the listener, is called in [buildHTTPProxyListener()](https://github.com/istio/istio/blob/bb93b2a69c73c0b3cbdb2f262e36a3e67bb369f5/pilot/pkg/networking/core/v1alpha3/listener_builder.go#L91). However, the `envoyFilterWrapper` is only populated in the [patchListener()](https://github.com/istio/istio/blob/bb93b2a69c73c0b3cbdb2f262e36a3e67bb369f5/pilot/pkg/networking/core/v1alpha3/listener_builder.go#L145), so the call to patchOneListener in buildHTTPProxyListener() do not apply any changes to the listener created for the HTTP Proxy.

So, to fix this bug, the patchOneListener for the HTTP Proxy listener moved to inside the patchListener(), where the other listener are patched. A new test were added to ensure that the listener are indeed removed.

Signed-off-by: Alan Diego dos Santos <alandiegosantos@gmail.com>
---
 .../core/v1alpha3/listener_builder.go         |  2 +-
 .../core/v1alpha3/listener_builder_test.go    | 39 +++++++++++++++++++
 2 files changed, 40 insertions(+), 1 deletion(-)

diff --git a/pilot/pkg/networking/core/v1alpha3/listener_builder.go b/pilot/pkg/networking/core/v1alpha3/listener_builder.go
index 82b61a88ed..09f07232dc 100644
--- a/pilot/pkg/networking/core/v1alpha3/listener_builder.go
+++ b/pilot/pkg/networking/core/v1alpha3/listener_builder.go
@@ -94,7 +94,6 @@ func (lb *ListenerBuilder) buildHTTPProxyListener() *ListenerBuilder {
 		return lb
 	}
 	removeListenerFilterTimeout([]*listener.Listener{httpProxy})
-	lb.patchOneListener(httpProxy, networking.EnvoyFilter_SIDECAR_OUTBOUND)
 	lb.httpProxyListener = httpProxy
 	return lb
 }
@@ -156,6 +155,7 @@ func (lb *ListenerBuilder) patchListeners() {
 
 	lb.virtualOutboundListener = lb.patchOneListener(lb.virtualOutboundListener, networking.EnvoyFilter_SIDECAR_OUTBOUND)
 	lb.virtualInboundListener = lb.patchOneListener(lb.virtualInboundListener, networking.EnvoyFilter_SIDECAR_INBOUND)
+	lb.httpProxyListener = lb.patchOneListener(lb.httpProxyListener, networking.EnvoyFilter_SIDECAR_OUTBOUND)
 	lb.inboundListeners = envoyfilter.ApplyListenerPatches(networking.EnvoyFilter_SIDECAR_INBOUND, lb.envoyFilterWrapper, lb.inboundListeners, false)
 	lb.outboundListeners = envoyfilter.ApplyListenerPatches(networking.EnvoyFilter_SIDECAR_OUTBOUND, lb.envoyFilterWrapper, lb.outboundListeners, false)
 }
diff --git a/pilot/pkg/networking/core/v1alpha3/listener_builder_test.go b/pilot/pkg/networking/core/v1alpha3/listener_builder_test.go
index 0a92983f7e..ce94d48d15 100644
--- a/pilot/pkg/networking/core/v1alpha3/listener_builder_test.go
+++ b/pilot/pkg/networking/core/v1alpha3/listener_builder_test.go
@@ -393,6 +393,45 @@ type fields struct {
 				},
 			},
 		},
+		{
+			name:  "remove HTTP Proxy listener",
+			proxy: sidecarProxy,
+			fields: fields{
+				httpProxyListener: &listener.Listener{
+					Name: "127.0.0.1_81",
+					Address: &core.Address{
+						Address: &core.Address_SocketAddress{
+							SocketAddress: &core.SocketAddress{
+								PortSpecifier: &core.SocketAddress_PortValue{
+									PortValue: 81,
+								},
+							},
+						},
+					},
+					FilterChains: []*listener.FilterChain{
+						{
+							Filters: []*listener.Filter{
+								{
+									Name: wellknown.HTTPConnectionManager,
+								},
+							},
+						},
+					},
+				},
+			},
+			want: fields{
+				inboundListeners: []*listener.Listener{
+					{
+						Name: "new-inbound-listener",
+					},
+				},
+				outboundListeners: []*listener.Listener{
+					{
+						Name: "new-outbound-listener",
+					},
+				},
+			},
+		},
 		{
 			name:  "patch add gateway listener",
 			proxy: gatewayProxy,
-- 
2.35.3

